### DISCUSSION TITLE
Help! Two extremely similar functions, yet one is forced to be `div`

### DISCUSSION THREAD

**Discussion by AZMCode** [koka-lang/koka#371] [Q&A] [answered]
I have been trying to create a small generator library for my own use, and I found that one piece of code I was writing, and encountering errors with, was very similar to an example within the documentation, yet wasn't working as well.

This is the code I've been writing:
```
pub effect<a> fun yield(item: a): ()

pub fun into-list<a,e>(gen: () -> <yield<a>|e> ()): e list<a>
  var l := []
  with handler
    fun yield(item) l := Cons(item,l)
    return(_) l.reverse
  gen()
 ```

This fails with the error message:
```
generator.kk(6,36): error: effects do not match
  context        :                                    l
  term           :                                    l
  inferred effect: <div|_e1>
  expected effect: $e
```
Adding the `div` effect both to the parameter and the return seems to work but..

A similar piece of code [copied straight from the documentation](https://koka-lang.github.io/koka/doc/book.html#sec-abstracting-handlers), works perfectly fine without it

```
effect fun emit( msg : string ) : ()

fun emit-collect( action : () -> <emit|e> () ) : e string
  var lines := []
  with handler
    return(x)     lines.reverse.join("\n") 
    fun emit(msg) lines := Cons(msg,lines)   
  action()
```

Why is the `div` effect being forced on my version of the code, but not on the documentation's?
The only perceivable difference I can think of is the fact that I'm returning a list and not a string, as well as the fact my effect is generic, but I don't see how either of those would make my function possibly diverge

**Comment by TimWhiting** [comment, marked as answer]
Divergence inference is rather conservative. Koka still has room for improvement in this area, and this particular issue seems worth looking into. However, annotating with divergence isn't the end of the world. In general it is hard(impossible) to determine if a program will ever diverge (the halting problem). That doesn't mean that for certain programs it can't be done, but that in general you will likely end up having to annotate divergence, even where you know as a programmer that it will never diverge. Program analyzers are getting better at figuring out some invariants similar to humans, but there is definitely more work to be done.

**Reply by AZMCode** [reply]
I see. I'd be interested in learning where the divergence inference breaks down exactly in this case. If you happen to have some pointers as to where I can learn the particulars of the divergence checker of the language, I'll be very thankful.

On another note, I understand annotating a function as "div" isn't the end of the world, but it does make it, for example, impossible to declare a constant whose value is calculated in terms of this function. Hope I am expressing myself clearly enough here.

**Reply by chtenb** [reply]
https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/paper-20.pdf Especially section 2.7

**Reply by TimWhiting** [reply]
@AZMCode 

You can use `unsafe-no-div(fn(){ divergent-function() })` if you know that something won't diverge to get rid of the divergence effect, if you want a top level variable expressed in terms of this.

**Reply by AZMCode** [reply]
Thanks for the tip!
I understand this can be done if it's absolutely necessary and no invariants are broken, but I also understand these kind of escape hatches to the protections set by the language should only be used as a last resort. Hopefully the divergence checker can detect these edge cases better.

**Reply by AZMCode** [reply]
@chtenb Thanks for the pointer!

**Reply by chtenb** [reply]
Or in some situations you can mark arguments to recursive calls as decreasing using `unsafe-decreasing`, which can be a clean way to hint the termination checker.

**Reply by AZMCode** [reply]
I'm guessing it doesn't apply in this exact situation though, as I can see no recursion happening

**Reply by chtenb** [reply]
Yeah, I meant in general
